vivarium/core/engine.py

```diff
@@ -563,32 +745,43 @@ def delete_path(
             if starts_with(path, deletion):
                 del self.process_paths[path]
 
-        for path in list(self.deriver_paths.keys()):
+        for path in list(self.step_paths):
             if starts_with(path, deletion):
-                del self.deriver_paths[path]
-
-    def run_derivers(self) -> None:
-        """Run all the derivers in the simulation."""
-        paths = list(self.deriver_paths.keys())
-        for path in paths:
-            # deriver could have been deleted by another deriver
-            deriver = self.deriver_paths.get(path)
-            if deriver:
-                # timestep shouldn't influence derivers
+                try:
+                    self.step_graph.remove(path)
+                except nx.exception.NetworkXError as e:
+                    # The step might have been deleted already.
+                    msg = f'The node {path} is not in the graph.'
+                    if e.args[0] != msg:
+                        raise e
+                del self.step_paths[path]
+
+    def run_steps(self) -> None:
+        '''Run all the steps in the simulation.'''
+        layers = self.step_graph.get_execution_layers()
+        for layer in layers:
+            deferred_updates: List[Tuple[Defer, Store]] = []
+            for path in layer:
+                step = self.step_paths[path]
+                if not step:
+                    # Step was deleted by a previous step.
+                    continue
+                # Timestep shouldn't influence steps.
                 # TODO(jerry): Do something cleaner than having
                 #  generate_paths() add a schema attribute to the Deriver.
                 #  PyCharm's type check reports:
```

Comment:
does this still apply?

Reply:
Yes, I kept this since I got errors without it